sips/committee_consensus.md

```diff
@@ -0,0 +1,480 @@
+type CommitteeRunner interface {
+	// Start the duty lifecycle for the given slot. Emits a message.
+    StartDuty(duty Duty) error
+	// Processes cosensus message
+    ProcessConsensus(consensusMessage *qbft.Message) error
+	// Processes a post-consensus message
+    ProcessPostConsensus(msg PartialSignatureMessages) 
+}
+
+// Committee runner implements the old runner interface
+type CommitteeRunner struct {
+	// Important fields only
+	Shares          map[ValidatorPubkey]Share
+    Operator        Operator
+	QBFTController *qbft.Controller
+	BeaconNetwork  *types.BeaconNetwork
+}
+
+// BeaconVote is the consensus data
+type BeaconVote struct {
+    BlockRoot         phase0.Root
+    Source            phase0.Checkpoint
+    Target            phase0.Checkpoint
+}
+
+func ConstructAttestation(vote BeaconVote, duty AttesterDuty) Attestation {
+    bits := bitfield.New(duty.CommitteeLength)
+    bits.Set(duty.ValidatorCommitteeIndex)
+
+    return Attestation{
+        Slot: duty.Slot,
+        BlockRoot: vote.BlockRoot,
+        Source: vote.Source,
+        Target: vote.Target,
+        CommiteeIndex: duty.CommitteeIndex,
+        AggregationBits: bits,
+    }
+}
+
+func ConstructSyncCommittee(vote BeaconVote, duty AttesterDuty) SyncCommitteeMessage{
+    return SyncCommitteeMessage
+    {
+        Slot: duty.Slot,
+        BlockRoot: vote.BlockRoot,
+        ValidatorIndex: duty.ValidatorIndex,
+    }
+}
+```
+#### PartialSignatureMessages
+
+The current structure that we have in code can be kept with small changes.
+
+```go
+type PartialSignatureMessages struct {
+	Type     PartialSigMsgType
+	Slot     phase0.Slot
+	Messages []*PartialSignatureMessage
+}
+
+// PartialSignatureMessage is a msg for partial Beacon chain related signatures (like partial attestation, block, randao sigs)
+type PartialSignatureMessage struct {
+	PartialSignature Signature `ssz-size:"96"` // The Beacon chain partial Signature for a duty
+	SigningRoot      [32]byte  `ssz-size:"32"` // the root signed in PartialSignature
+	Signer         OperatorID
+	ValidatorIndex phase0.ValidatorIndex	// addition
+}
+```
+
+We must have the following flow:
+1. Calculate `expectedRootsAndBeaconObjects()`. Use the `duty` objects to reconstruct the proper beacon data objects (i.e. `Attestation`).
+2. The above calculation can be used to create a mapping of `ValidatorIndex` to `root`, and `root` to `beaconObject`.
+3. When processing the messages find all the roots that have quorums for a certain validator, mark them, and submit corresponding beacon data to beacon chain.
+4. For the next message received attempt to complete quorum for other roots.
+
+
+#### Happy Flow
+
+1. `Committee` receives duties that match a certain slot. Then `StartDuty` and filter the validators who have a higher `HighestAttestingSlot`. Initialize a `CommitteeRunner` for the relevant Validators and slot. For each validator that we start, mark its old beacon duties as stopped.
```

Comment:
I don't see it as a security concern as slashing protection protects against it. Should be on implementation side

Reply:
ok

Reply:
we need to make sure that in spec we have slashing check enabled before we sign partial sigs even if they pass consensus